### ISSUE TITLE
Segfault encountered while analyzing on macOS

### ISSUE THREAD

**Issue by shyndman** [dart-lang/sdk#40809]
Hi there,

While using the analyzer in a release mode Flutter project, I encountered a segfault. I rebuilt in debug mode, got into the same state, and received an exception where the segfault was encountered previously. 

Turns out that the file I was analyzing was referencing a package that was not referenced in the project's pubspec. Adding the dep fixed the issue.

Here's the segfault info: https://gist.github.com/shyndman/54e43ed87be13d0822619959f6a3f5bb

And here's the exception (there were a few, but the stack seems to resemble the segfault's):
```
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:53:5)
#1      FeatureSetProvider.getFeatureSet (package:analyzer/src/dart/analysis/feature_set_provider.dart:30:13)
#2      FileState._parse (package:analyzer/src/dart/analysis/file_state.dart:571:50)
#3      FileState.parse.<anonymous closure> (package:analyzer/src/dart/analysis/file_state.dart:368:16)
#4      _PerformanceTagImpl.makeCurrentWhile (package:analyzer/src/generated/utilities_general.dart:256:15)
#5      FileState.parse (package:analyzer/src/dart/analysis/file_state.dart:367:42)
#6      FileState.refresh (package:analyzer/src/dart/analysis/file_state.dart:419:32)
#7      FileSystemState.unresolvedFile (package:analyzer/src/dart/analysis/file_state.dart:823:23)
#8      FileSystemState.getFileForUri (package:analyzer/src/dart/analysis/file_state.dart:884:27)
#9      FileState._fileForRelativeUri (package:analyzer/src/dart/analysis/file_state.dart:548:21)
#10     FileState.refresh (package:analyzer/src/dart/analysis/file_state.dart:476:18)
#11     FileSystemState.getFileForPath (package:analyzer/src/dart/analysis/file_state.dart:854:12)
#12     AnalysisDriver._computeAnalysisResult (package:analyzer/src/dart/analysis/driver.dart:1219:31)
#13     AnalysisDriver.performWork (package:analyzer/src/dart/analysis/driver.dart:932:33)
<asynchronous suspension>
#14     AnalysisDriverScheduler._run (package:analyzer/src/dart/analysis/driver.dart:1979:24)
<asynchronous suspension>
#15     AnalysisDriverScheduler.start (package:analyzer/src/dart/analysis/driver.dart:1913:5)
#16     ContextBuilderImpl.createContext (package:analyzer/src/dart/analysis/context_builder.dart:75:17)
#17     new AnalysisContextCollectionImpl
```

**Comment by mraleph** [comment]
Seems like a null reference causing a segfault potentially? /cc @mkustermann @alexmarkov 

We will take a look - but for us to really zero on it we would need a reproduction.

**Comment by shyndman** [comment]
Whew, took me a bit, but I got it.
https://github.com/shyndman/analyzer_segfault_repro

**Comment by shyndman** [comment]
And after diving in with the debugger, I _think_ that this line is the culprit:
https://github.com/dart-lang/sdk/blob/4dad6d77ba50e0440689a5a7c8ba7c77915a4aa2/pkg/analyzer/lib/src/dart/analysis/feature_set_provider.dart#L33

In the case of `FileStateSystem`'s unresolved file, `uri` is `null`.

That said, the code is written to handle an exception here reasonably. There's a catch higher up in the stack that has reasonable default behavior. 🤷‍♂

**Comment by mraleph** [comment]
Thank you for the reproduction @shyndman, fix is under review here: https://dart-review.googlesource.com/c/sdk/+/137965

**Comment by shyndman** [comment]
Thank you.

Not to burden you, but would you mind explaining a little more about the fix, for a compiler laymen. It looks interesting. 

What's a catch entry? What does it mean to move it?

**Comment by mraleph** [comment]
In layman terms: every time an exception is caught by a catch block we need to ensure that values of the variables are stored on the stack in locations where catch block expects them. During try block they might be stored in completely different places (as compiler sees fit to place them). Thus some reshuffling is needed - this reshuffling is described by _catch entry moves_, a list of moves which describe what needs to be moved where. The bug was that we did not move `this` variable to the correct location if it was captured by an inner closure. 

You can try looking through https://github.com/dart-lang/sdk/blob/master/runtime/docs/compiler/exceptions.md if you are curious - though it requires some familiarity with things like SSA form to grok.

**Comment by shyndman** [comment]
Thank you @mraleph. Interesting reading. :)
